### PR TITLE
[15.0][ADD] stock_inventory

### PR DESCRIPTION
**Pull request by DavidJForgeFlow**
This module allows to group Inventory Adjustments and have a group traceability (like before 15.0).

@ForgeFlow

### CODE REVIEW

**Comment by rousseldenis** [comment]
@DavidJForgeFlow Thanks for this. @pedrobaeza I think you could look at it following what we discussed on OCA days.

**Comment by pedrobaeza** [comment]
Yeah, I see.

David, please name the module `stock_inventory_adjustment` (without `s` and I suppose you wanted to say "grouped inventory", but I think that having "inventory adjustment" in the name is enough.

File: stock_group_inventory_adjustments/models/stock_adjustment.py
Comment (pedrobaeza) [inline stock_group_inventory_adjustments/models/stock_adjustment.py]: Can't this still be called `stock.inventory` for avoiding migration scripts?

Comment (rousseldenis) [reply stock_group_inventory_adjustments/models/stock_adjustment.py]: Maybe we can have a feedback by someone having done Odoo migrations to 15.0 to see what they have done with those records.

**Comment by DavidJForgeFlow** [comment]
> Yeah, I see.
> 
> David, please name the module `stock_inventory_adjustment` (without `s` and I suppose you wanted to say "grouped inventory", but I think that having "inventory adjustment" in the name is enough.

All references changed. Also changed the model name to stock.inventory, is it ok like this?

File: stock_inventory_adjustment/__manifest__.py
Comment (GuillemCForgeFlow) [inline stock_inventory_adjustment/__manifest__.py]:
```suggestion
    "version": "15.0.1.0.0",
```

File: stock_inventory_adjustment/views/stock_inventory.xml
Comment (GuillemCForgeFlow) [inline stock_inventory_adjustment/views/stock_inventory.xml]:
```suggestion
                            attrs="{'invisible':['|', ('state', 'in', ['2', '3']), ('count_stock_quants', '=', 0)]}"
                        >
                            <field name="count_stock_quants" readonly="1" string="Adjustments" />
```
For all smart buttons:
- Move the string to the field
- Also, make it invisible if the count is 0.

Comment (DavidJForgeFlow) [reply stock_inventory_adjustment/views/stock_inventory.xml]: Done!

Comment (GuillemCForgeFlow) [reply stock_inventory_adjustment/views/stock_inventory.xml]: Great :+1: . I will wait for the other changes then.

File: stock_inventory_adjustment/models/stock_inventory.py
Comment (rousseldenis) [inline stock_inventory_adjustment/models/stock_inventory.py]: I don't like the use of integer as state value as looking at this code, you cannot determine what state is behind

File: stock_inventory_adjustment/models/stock_inventory.py
Comment (rousseldenis) [inline stock_inventory_adjustment/models/stock_inventory.py]: Please use:
```suggestion
            self.stock_quant_ids : self.env["stock.quant"].search([("location_id", "in", self._origin.location_ids.ids)])
            
```

Comment (DavidJForgeFlow) [reply stock_inventory_adjustment/models/stock_inventory.py]: Done! thank you!

File: stock_inventory_adjustment/models/stock_inventory.py
Comment (DavidJForgeFlow) [reply stock_inventory_adjustment/models/stock_inventory.py]: Also done, thank you again!

File: stock_inventory_adjustment/views/stock_inventory.xml
Comment (DavidJForgeFlow) [reply stock_inventory_adjustment/views/stock_inventory.xml]: You changes are done :smile:

**Review by GuillemCForgeFlow — APPROVED** [review]
LGTM! :top: :+1:

File: stock_inventory/models/stock_inventory.py
Comment (CasVissers-360ERP) [inline stock_inventory/models/stock_inventory.py]: String = "Lot" would be cleaner here?

File: stock_inventory/models/stock_inventory.py
Comment (CasVissers-360ERP) [inline stock_inventory/models/stock_inventory.py]: String = "Lots" would be cleaner here?

File: stock_inventory/models/stock_inventory.py
Comment (CasVissers-360ERP) [inline stock_inventory/models/stock_inventory.py:159]: Should we extend this? Checking if there are counts with same product or same lots or same locations? Counting two locations at the same time should be possible?

File: stock_inventory/models/stock_inventory.py
Comment (CasVissers-360ERP) [inline stock_inventory/models/stock_inventory.py]: It is still a manual action to confirm the quant count?

**Review by CasVissers-360ERP — CHANGES_REQUESTED** [review]
Functional review

File: stock_inventory/models/stock_inventory.py
Comment (rousseldenis) [reply stock_inventory/models/stock_inventory.py]: I would say `Lot/Serial Number` to use Odoo naming

Comment (rousseldenis) [reply stock_inventory/models/stock_inventory.py]: @DavidJForgeFlow Maybe an 'owner' management too ?

Comment (CasVissers-360ERP) [reply stock_inventory/models/stock_inventory.py]: I agree with Lot/Serial Number

File: stock_inventory/models/stock_inventory.py
Comment (rousseldenis) [inline stock_inventory/models/stock_inventory.py]: I would have put all domains in individual methods in order to allow override

File: stock_inventory/models/stock_inventory.py
Comment (CasVissers-360ERP) [reply stock_inventory/models/stock_inventory.py]: Edit: Lot/Serial Numbers

File: stock_inventory/models/stock_inventory.py
Comment (rousseldenis) [inline stock_inventory/models/stock_inventory.py]: Better : 

```suggestion
        self.stock_quant_ids.update({"to_do": True})
```

File: stock_inventory/models/stock_inventory.py
Comment (DavidJForgeFlow) [reply stock_inventory/models/stock_inventory.py]: @rousseldenis what do you mean with 'owner'? Add a responsible field?
The other changes are done, thanks!

File: stock_inventory/models/stock_inventory.py
Comment (DavidJForgeFlow) [reply stock_inventory/models/stock_inventory.py]: Done thanks!

File: stock_inventory/models/stock_inventory.py
Comment (DavidJForgeFlow) [reply stock_inventory/models/stock_inventory.py:159]: I don't think so because you can only do an Adjustment at a time, so repeated counts are "imposible" (unless something magic happens).

File: stock_inventory/models/stock_inventory.py
Comment (DavidJForgeFlow) [reply stock_inventory/models/stock_inventory.py]: Yes, confirming a count is still manual.

File: stock_inventory/models/stock_inventory.py
Comment (DavidJForgeFlow) [reply stock_inventory/models/stock_inventory.py]: Done, thanks!

File: stock_inventory/models/stock_inventory.py
Comment (DavidJForgeFlow) [reply stock_inventory/models/stock_inventory.py]: Done!

File: stock_inventory/models/stock_inventory.py
Comment (CasVissers-360ERP) [reply stock_inventory/models/stock_inventory.py]: Ok, this is very confusing? Should we at least give a warning to the user that the count(s) still needs to be confirmed? Setting a group to done for me means that the count is confirmed.

Comment (CasVissers-360ERP) [reply stock_inventory/models/stock_inventory.py]: What if a quant is being mutated which is currently in a stock adjustment which is not done yet, should we not block this? Same like not allowing 2 counts? Overall I have a sense the module is confusing now. Why do we add a state to stock_inventory if nothing is being done / blocked or checked? If it is simply a group of quants, then we can create a group without a workflow?

File: stock_inventory/models/stock_inventory.py
Comment (rousseldenis) [inline stock_inventory/models/stock_inventory.py]:
```suggestion
        required=True, default="Inventory", string="Inventory Reference"
```

File: stock_inventory/models/stock_inventory.py
Comment (rousseldenis) [reply stock_inventory/models/stock_inventory.py]: This one : https://github.com/odoo/odoo/blob/15.0/addons/stock/models/stock_quant.py#L76

File: stock_inventory/models/stock_inventory.py
Comment (rousseldenis) [inline stock_inventory/models/stock_inventory.py]: Maybe you could put domains retrievals in separate methods to easy possible inheritance?

Comment (DavidJForgeFlow) [reply stock_inventory/models/stock_inventory.py]: All the different type of Inventory Adjustments are computed in different methods, what do you mean with putting the domains retrievals in separate methods?

File: stock_inventory/models/stock_inventory.py
Comment (DavidJForgeFlow) [reply stock_inventory/models/stock_inventory.py]: Done, thank you!

**Review by AaronHForgeFlow — COMMENTED** [review]
Works Good!

Just a minor thing that needs to be fixed:

* Create an inventory adjustment for one product only.
* Start the adjustment
* Change the quantity and click on Apply
* The stock move line is successfully linked
* Go to the stock.inventory again and then to the adjustments, remove the filter "Todo".
* Change the quantity again to a new one and click on apply. 
* Issue: A new quant is successfully created but it is not linked to the inventory adjustment, that is, from the button Stock move lines I can only see one of the stock moves lines

**Comment by DavidJForgeFlow** [comment]
> Works Good!
> 
> Just a minor thing that needs to be fixed:
> 
>     * Create an inventory adjustment for one product only.
> 
>     * Start the adjustment
> 
>     * Change the quantity and click on Apply
> 
>     * The stock move line is successfully linked
> 
>     * Go to the stock.inventory again and then to the adjustments, remove the filter "Todo".
> 
>     * Change the quantity again to a new one and click on apply.
> 
>     * Issue: A new quant is successfully created but it is not linked to the inventory adjustment, that is, from the button Stock move lines I can only see one of the stock moves lines

Thank you Aaron, this is already fixed!

The code is now better to inherit because of some change of how domains work, thanks to @GuillemCForgeFlow and @rousseldenis for the idea. 

Also, added the possiblilty of doing two Adjustments on two different Locations at the same time.

**Comment by GuillemCForgeFlow** [comment]
Thanks for the work @DavidJForgeFlow :+1: !

**Review by AaronHForgeFlow — APPROVED** [review]
Functional review LGTM!

**Comment by CasVissers-360ERP** [comment]
@AaronHForgeFlow If you did a functional approval I am really curious about feedback on my questions. For me this module currently makes no sense. It is just a group of quants and could better be called stock_quant_group.

**Comment by AaronHForgeFlow** [comment]
Hi @CasVissers-360ERP  It is called stock_inventory in order to respect what it was in standard in previous versions, otherwise the migration will get more complex.

I guess the main discussion regards this comment:

> What if a quant is being mutated which is currently in a stock adjustment which is not done yet, should we not block this? Same like not allowing 2 counts? Overall I have a sense the module is confusing now. Why do we add a state to stock_inventory if nothing is being done / blocked or checked? If it is simply a group of quants, then we can create a group without a workflow?

I think that if the quant is altered while the inventory adjustment is done then there is no inconsistencies in the database in this case, because Odoo will internally handle this. It may be confusing for users that quants did a change while they were doing the adjustments.  In that case maybe it is necessary to avoid the modification of the quant until the inventory adjustment is completed. But that can be done on a separate module I think. I think there was also a module for that in this repo.

**Comment by CasVissers-360ERP** [comment]
@AaronHForgeFlow 

You only answered one of my concerns? My main concern: Why do we add a state to stock_inventory if nothing is being done / blocked or checked? In addition it is not an inventory / inventory count. An inventory count start which blocking the location, counting the stock, and adjusting the stock level accordingly. I miss the whole idea of creating a count in Odoo (group of counts). But if there is no logic in the group/inventory I feel like it will be very confusing for the end-user. I have a hard time understanding the benefits as well. 

If the goal of this module is to simply create a group of quants, just remove the state and rename the module.

**Comment by AaronHForgeFlow** [comment]
Hi @CasVissers-360ERP ,  I am just saying that the module works as my customer needs. They just need a document that records and keep a history of the changes on inventory. They are not worried about that restriction because the lifecycle of an inventory adjustment is too short for them and that kind of "issues" are not relevant. Nothing is blocked the quant in this module but it can be blocked on a module that depends on this one. Perhaps that feature can be included here, I am not against that.

**Comment by CasVissers-360ERP** [comment]
@AaronHForgeFlow Well maybe the module is not generic/mature enough to be on OCA module?

**Comment by AaronHForgeFlow** [comment]
@CasVissers-360ERP I think it is very generic module, it recovers the standard inventory adjustments feature form previous versions and that is proved to be generic enough. The only thing that may be missing is the locking on the quant that is being inspected. It is ok to me to implement that restriction on this module, but I will not block this PR because of that. Would you be fine with this PR if it implements that restriction?

Those discussions are good for the quality of the module so I am glad you commented and I hope more people joins to the discussion here.

**Review by ventor-dev — APPROVED** [review]
Great work! Customers still need this functionality to be able to group inventories for individual products into single big document. That is much better for Warehouse manager to review it later
 Not sure I understand why Odoo removed it. 

@DavidJForgeFlow thank you for the work! Hope it will be merged soon into OCA. Approved from my side

**Comment by CasVissers-360ERP** [comment]
@AaronHForgeFlow I don't agree with this " it recovers the standard inventory adjustments "

A big thing missing is:
https://github.com/odoo/odoo/blob/e9a7283ea33a5e122afa53f73f717ed45ce9874d/addons/stock/models/stock_inventory.py#L131

But since everybody else seems to approve this PR I think it's best for me to exit the discussion.

**Comment by LoisRForgeFlow** [comment]
Obviously this can receive improvements and changes, but since I think it is a common requirement for customers coming from previous Odoo versions, let's at least cover the gap with this (I guess better this than nothing).

Thanks to all participants! Do not hesitate to open PRs to include improvements and changes once this gets merged.

/ocabot merge nobump

**Pull request merged**
